---
title: "Using BiocProject with tximeta"
author: "MichaÅ‚ Stolarczyk"
date: "2020-10-28"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Using BiocProject with tximeta"}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


Before you start see the 
[Getting started with `BiocProject` vignette](./vignette1getStarted.html) for the basic information and installation instructions.

# Introduction

This vignette demonstrates how to integrate BiocProject with [tximeta Bioconductor package](https://www.bioconductor.org/packages/release/bioc/html/tximeta.html) for a really slick start-to-finish analysis of RNA-seq data.

Tximeta is a tool that automatically annotates RNA-seq samples during the import of transcript quantification files from the [salmon](https://salmon.readthedocs.io/en/latest/salmon.html) transcript quantifier. How it works, is that `salmon` records a unique identifier of the transcriptome it uses during quantification; when importing the transcript values into R, tximeta reads this identifier and looks up metadata about those sequences using a local database of known transcriptome identifiers. For more details, refer to the [tximeta GitHub repository](https://github.com/mikelove/tximeta) or [publication in PLoS Computational Biology](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007664).

The `tximeta::tximeta` function takes as input a `data.frame` (`coldata`) object that, for Salmon results, points to quantification results directory for each sample. The `tximeta` function reads the `*.sa` files and returns a single `SummarizedExperiment` object with the Salmon-generated metadata in the object `metadata` slot.

Since `SummarizedExperiment` inherits from `Annotated` class, it fits perfectly into `BiocProject` output object class requirements. 



```r
suppressPackageStartupMessages(library(BiocProject))
suppressPackageStartupMessages(library(SummarizedExperiment))
is(SummarizedExperiment(), "Annotated")
```

```
## [1] TRUE
```

## Advantages of using BiocProject with tximeta

If we add BiocProject in to the tximeta workflow, then sample metadata from the PEP project specification can be easily plugged in to the tximport workflow. For example, if a researcher used a PEP to run Salmon to quantify reads across multiple samples with PEP-compatible workflow management engine/job scatterer like [Snakemake](https://snakemake.github.io/), [CWL](https://www.commonwl.org/) or [looper](https://looper.databio.org/), the same PEP would be ready to use with tximeta as long as the samples had `files` attribute defined. This could be done either via a `files` column in the sample table, or by using one of the sample modifiers provided by the PEP framework. The advantages of calling `tximport` within `BiocProject` include: 

 - project portability, inherent to projects following PEP specification
 - single source of metadata from start of the analysis to finish -- all the PEP-defined metadata will be propagated to the output object of `tximeta` function. It will be accessible with [pepr](http://code.databio.org/pepr/) API or `@PEP` in the `metadata` slot of `SummarizedExperiment`, just as any other metadata attached to the result by `tximeta` function.

# Setup

## PEP

In order to use `tximeta` and `Biocproject` functions, we need a PEP. Let's use one that looks like this:


```
  Error in file(file, "rt", encoding = fileEncoding): cannot open the connection
```





















