---
title: "Using BiocProject with tximeta"
author: "MichaÅ‚ Stolarczyk"
date: "`r Sys.Date()`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Using BiocProject with tximeta"}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r, echo=FALSE}
knitr::opts_chunk$set(collapse=FALSE, message=FALSE)
```

# Introduction

## Prerequisites

This vignette demonstrates how to integrate BiocProject with the [tximeta Bioconductor package](https://www.bioconductor.org/packages/release/bioc/html/tximeta.html) for a really slick start-to-finish analysis of RNA-seq data. We assume you're familiar with BiocProject; if not, please start with [Getting started with `BiocProject` vignette](./vignette1getStarted.html) for basic instructions.

## Introduction to Tximeta

Tximeta is a package that imports transcript quantification files from the [salmon](https://salmon.readthedocs.io/en/latest/salmon.html) transcript quantifier. When importing, tximeta automatically annotates the data with the transcriptome used. How it works is that `salmon` records a unique identifier of the transcriptome it uses during quantification; then, tximeta reads this identifier and looks up metadata about those sequences using a local database of known transcriptome identifiers. For more details, refer to the [tximeta GitHub repository](https://github.com/mikelove/tximeta) or [publication in PLoS Computational Biology](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007664).

The `tximeta::tximeta` function takes as input a `data.frame` (`coldata`) object that, for Salmon results, points to a quantification results directory for each sample. The `tximeta` function reads the `*.sa` files and returns a single `SummarizedExperiment` object with the Salmon-generated metadata in the object `metadata` slot.

Since `SummarizedExperiment` inherits from the Bioconductor `Annotated` class, it fits perfectly into `BiocProject` output object class requirements. 


```{r}
suppressPackageStartupMessages(library(BiocProject))
suppressPackageStartupMessages(library(SummarizedExperiment))
is(SummarizedExperiment(), "Annotated")
```

## Advantages of using BiocProject with tximeta

If we add BiocProject in to the tximeta workflow, then sample metadata from the PEP project specification can be easily plugged in! For example, if a researcher used a PEP to run Salmon to quantify reads across multiple samples with PEP-compatible workflow management engine/job scatterer like [Snakemake](https://snakemake.github.io/), [CWL](https://www.commonwl.org/), or [looper](https://looper.databio.org/), the same PEP would be ready to use with tximeta as long as the samples had `files` attribute defined. This could be done either via a `files` column in the sample table, or by using one of the sample modifiers provided by the PEP framework. The advantages of calling `tximport` within `BiocProject` include: 

 - project portability, inherent to projects following PEP specification
 - single source of metadata from start of the analysis to finish -- all the PEP-defined metadata will be propagated to the output object of the `tximeta` function automatically. It will be accessible from within your R session using the [pepr](http://code.databio.org/pepr/) API, or with `@PEP` in the `metadata` slot of the `SummarizedExperiment` object, just as any other metadata attached to the result by `tximeta` function.

Let's show you how this work with a simple demo.

# Demo of the BiocProject + tximeta workflow

## Download example data

First, let's download some RNA-seq counts from salmon, described in PEP format:

```{r, download-data, collapse=TRUE, comment=" "}
if (basename(getwd()) != "long_vignettes") setwd("long_vignettes")
pth = BiocFileCache::bfcrpath(
  BiocFileCache::BiocFileCache(getwd()), 
  "http://big.databio.org/example_data/tximeta_pep.tar.gz"
  )
utils::untar(tarfile=pth)
abs_pep_path = file.path(getwd(), "tximeta_pep")
abs_cfg_path = file.path(abs_pep_path, "project_config.yaml")
```

Let's take a look at what we have here...

## Examine and load the PEP into R

The `Biocproject` + `tximeta` workflow requires a PEP. The example we just downloaded looks like this:

```{r, warning=FALSE, echo=FALSE, message=FALSE, collapse=TRUE, comment=" "}
library(pepr)
.printNestedList(yaml::read_yaml(abs_cfg_path))
```

As you can see, this PEP configuration file uses a `$TXIMPORTDATA` environment variable to specify a file path. This is just an optional way to make this PEP work in any computing environment without being changed, so you can share your sample metadata more easily. For this vignette, we need to set the variable to the output directory where our downloaded results are stored:

```{r}
Sys.setenv("TXIMPORTDATA"=file.path(abs_pep_path, "/tximportData"))
```

```{r eval=TRUE, include=FALSE}
# Run some stuff we need for the vignette
p=Project(abs_cfg_path)
```

Now, look at the `sample_table` key in the configuration file. It points to the second major part of a PEP: the 
sample table CSV file (``r { basename(config(p)$sample_table) }``). Check out the contents of that file:

```{r, echo=FALSE, message=FALSE, warning=FALSE, collapse=TRUE, comment=" "}
library(knitr)
coldataDF = read.table(p@config$sample_table, sep=",", header=TRUE)
knitr::kable(coldataDF, format = "html")
```

This sample table lacks the `files` column required by tximeta -- but this file is sufficient, since BiocProject, or more specifically pepr, will take care of constructing the portable `files` sample attribute automatically via `sample_modifiers.derive`, where the config file above specifies the `files` attribute and its path. 

Now we can load the file with BiocProject... but first, a short detour

## Detour: the magic of PEP sample modifiers 

Before we jump into using `BiocProject`, let's take a minute to demonstrate how using the PEP helps us out here. Let's read in our PEP using the the generic `Project` function from `pepr`:


```{r}
p=Project(abs_cfg_path)
```

We now have our PEP project read in, and we can see what is found in the sample table:

```{r}
sampleTable(p)
```

See how our sample table has now been automatically updated with the `files` attribute? *That* is the magic of the PEP sample modifiers. It's that simple. Now, let's move on to demonstrate what `BiocProject` adds.

## The BiocProject data processing function

If you look again at our configuration file above, you'll notice the `biconductor` section in the configuration file, which defines a function name and R script. These specify the BiocProject data processing function, which in this case, is simply a `tximeta` call that uses the PEP-managed processed sample table its input. Here's what that function looks like:

```{r echo=FALSE, eval=TRUE, comment=""}
source(file.path(abs_pep_path, "readTximeta.R"))
get(config(p)$bioconductor$readFunName)
```

## Loading in the data with BiocProject

We have everything we need: a salmon output file, a PEP that specifies a sample table and provides the `files` column, and a function that uses `tximeta` to create the final `SummarizedExperiment` object. Now, we can call the `BiocProject` function:

```{r collapse=TRUE}
require(tximeta)
bp = BiocProject(abs_cfg_path)
```

The output of `BiocProject` function, the `bp` object in our case, is magical. In one object, it supports the functionality of `SummarizedExperiment`, `tximeta`, and `pepr`. Observe:

First, it is a `RangedSummarizedExperiment`, so it supports all methods defined in `SummarizedExperiment`:

```{r}
suppressPackageStartupMessages(library(SummarizedExperiment))
colData(bp)
assayNames(bp)
rowRanges(bp)
```

Naturally, we can use tximeta methods:

```{r collapse=TRUE}
retrieveDb(bp)
```

But wait, there's more! The `PEP` metadata information has been attached to the metadata as well. Let's extract the `Project` object from the result with `getProject` method:

```{r collapse=TRUE}
getProject(bp)
```

You can use the `pepr` API for any R-based PEP processing tools:

```{r collapse=TRUE}
sampleTable(bp)
config(bp)
```

# Conclusion

If you format your project metadata according to the PEP specification, it will be ready to use with tximeta and the resulting object will include project-wide metadata and expose [pepr](http://code.databio.org/pepr/) API for any PEP-compatible R packages for downstream analysis.
