---
title: "Using BiocProject with tximeta"
author: "MichaÅ‚ Stolarczyk"
date: "`r Sys.Date()`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Using BiocProject with tximeta"}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r, echo=FALSE}
knitr::opts_chunk$set(collapse=FALSE, message=FALSE)
```

# Introduction

Another example of BiocProject usage is its seamless integration with [tximeta Bioconductor package](https://www.bioconductor.org/packages/release/bioc/html/tximeta.html).

In brief, tximeta is a tool that performs numerous annotation and metadata gathering tasks automatically during the import of transcript quantification files, e.g., result of [Salmon](https://salmon.readthedocs.io/en/latest/salmon.html) transcript quantifier. Tximeta provides reference provenance information by storing a local database of known transcriptome checksums. For more details regarding the tool please refer to:

- [GitHub repository](https://github.com/mikelove/tximeta)
- [publication](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007664)

Importantly, the required input to the `tximeta::tximeta` function is a `data.frame` (`coldata`) object that, in case of Salmon results processing, points to quantification results directory for each sample. `*.sa` files are read and saved as a single `SummarizedExperiment` object with the Salmon-generated metadata in the object `metadata` slot.

## Advantages of the integration

Since `SummarizedExperiment` inherits from `Annotated` class, it fits perfectly into `BiocProject` output object class requirements. Therefore, sample metadata being part of a project formatted according to the PEP specification can be easily plugged in to the tximport workflow. For example, if a researcher used PEP to run Salmon to quantify reads across multiple samples with PEP-compatible workflow management engine/job scatterer like [Snakemake](https://snakemake.github.io/), [CWL](https://www.commonwl.org/) or [looper](https://looper.databio.org/) the same PEP would be ready to use with tximeta as long as the samples have `files` attribute defined (via `files` column in the sample table or via one of the sample modifiers in the PEP framework).

```{r}
suppressPackageStartupMessages(library(BiocProject))
suppressPackageStartupMessages(library(SummarizedExperiment))
is(SummarizedExperiment(), "Annotated")
```

# Setup

## PEP

In order to use `tximeta` and `Biocproject` functions, we need a PEP. Let's use one that looks like this:

```{r, warning=FALSE, echo=FALSE, message=FALSE, collapse=TRUE, comment=" "}
library(pepr)
configFile = "/Users/mstolarczyk/Desktop/tximeta-pep/project_config.yaml"
.printNestedList(yaml::read_yaml(configFile))
```

```{r eval=TRUE, include=FALSE}
# Run some stuff we need for the vignette
p=Project(configFile)
readTximeta = "/Users/mstolarczyk/Desktop/tximeta-pep/readTximeta.R"
source(readTximeta)
```

As you can see in the PEP configuration file above `$TXIMPORTDATA` environment variable has been used to make the path portable. Therefore, we need to set it. A reduced version of tximportData package contents is shipped with BiocProject, so let's use this sample dataset:

```{r}
tximportDataPath = "/Users/mstolarczyk/Desktop/tximeta-pep/tximportData"
Sys.setenv("TXIMPORTDATA"=tximportDataPath)
```

This configuration file points to the second major part of a PEP: the 
sample table CSV file (``r { basename(config(p)$sample_table) }``). Check out the contents of that file:

```{r, echo=FALSE, message=FALSE, warning=FALSE, collapse=TRUE, comment=" "}
library(knitr)
coldata = "/Users/mstolarczyk/Desktop/tximeta-pep/sample_table.csv"
coldataDF = read.table(coldata, sep=",", header=TRUE)
knitr::kable(coldataDF, format = "html")
```

Even though the required `files` column is still missing this file is sufficient, since BiocProject, or more specifically pepr, will take care of constructing the portable `files` sample attribute automatically via `sample_modifiers.derive`:

```{r}
p=Project(configFile)
sampleTable(p)
```

*Note that the use of `Project` function in the chunk above is not required to complete this tutorial. It is used solely to present the effect of PEP metadata processing*

## Data processing function

Another required component is the data processing function, which is simply a `tximeta` call that uses the PEP-managed processed sample table its input:

```{r eval=FALSE}
function (pep) 
{
    require(tximeta)
    tximeta::tximeta(coldata = sampleTable(pep))
}
```
However, following tximeta instructions, to avoid downloading remote GTF files during this vignette, we will point to a GTF file saved locally (in the tximportData package). We link the transcriptome of the Salmon index to its locally saved GTF. The standard recommended usage of `tximeta` would be the code chunk above, or to specify a remote GTF source, not a local one. This following code is therefore not recommended for a typically workflow, but is particular to the vignette code.
```{r echo=FALSE, eval=TRUE, comment=""}
get(config(p)$bioconductor$readFunName)
```

# Usage

Once we have the PEP and a function that will be used to create the final `SummarizedExperiment` object we can call `BiocProject` function:

```{r collapse=TRUE}
require(tximeta)
bp = BiocProject(configFile)
```
After successful import we can browse the resulting object. Since it is a `RangedSummarizedExperiment`, the methods defined in SummarizedExperiment package work:
```{r}
suppressPackageStartupMessages(library(SummarizedExperiment))
colData(bp)
assayNames(bp)
rowRanges(bp)
```
Naturally, we can use tximeta methods:
```{r collapse=TRUE}
retrieveDb(bp)
```
And finally, note that the `PEP` metadata information has been attached to the metadata as well. Let's extract the `Project` object from the result with `getProject` method from this package:

```{r collapse=TRUE}
getProject(bp)
```